Add priority levels to production exits

pchote opened the pull request:
This PR adds the ability to specify multiple priorities for exits, with lower priorities only being considered if all higher priority exits are blocked.

This is used to fix the exit behaviour for TS barracks: each defines two normal exits (0,2 and 1,2 for GDI; 2,2 and 3,2 for Nod) and all the remaining perimeter cells as fallback.

<img width="576" alt="Screenshot 2019-10-05 at 17 13 25" src="https://user-images.githubusercontent.com/167819/66257700-59d7c600-e794-11e9-90ae-e81bd65b6954.png">

Fixes #10843.
Supersedes #17105.

pchote committed:
Add priority levels to Exit.

pchote committed:
Add fallback exits to TS barracks.

Mailaender commented on OpenRA.Mods.Common/Traits/Buildings/RallyPoint.cs:
```diff
@@ -70,7 +70,13 @@ public RallyPoint(Actor self, RallyPointInfo info)
+				.GroupBy(e => e.Priority)
+				.FirstOrDefault();
+
+			var exits = priorityExits != null ? priorityExits.ToArray() : new ExitInfo[0];
```
What does the `new ExitInfo[0]` do?

tovl reviewed (changes requested):
It would be a good idea to add fallback exits to RA kennels as well, so we can close #16792.

matjaeck commented:
Purist comment:
![kennel](https://user-images.githubusercontent.com/37185497/69184544-59617780-0b15-11ea-990a-b99eb8c04689.gif)

matjaeck commented:
FWIW, my comment above was not meant to be an objection against adding fallback exits to the kennel. It's probably good enough.

pchote committed:
Add fallback exits to RA Kennel.

pchote commented:
Rebased and updated Kennel exits:

<img width="346" alt="Screenshot 2019-12-01 at 00 28 44" src="https://user-images.githubusercontent.com/167819/69907715-dcb28100-13d1-11ea-80cb-de456cfd09e8.png">

The primary exit is to the bottom-left (-1,1). If this is blocked it will prefer (-1,0) and (0,1) before falling back again to the other 5 cells.

tovl reviewed (approved):
It might look better if the spawn point for the kennel exits (0, -1), (1, -1) and (1, 0) was placed more to the upper right. This is not a blocker though.

tovl commented:
@matjaeck @teinarss do you consider your +1 still valid?

teinarss commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/707636bce28b633585913c730b27bb3c68782095...60b6ac76d199e42fea75ef59a888bebd73d7500f
